sdk/tests/deqp/functional/gles3/es3fOcclusionQueryTests.js

```diff
@@ -0,0 +1,461 @@
+
+                bufferedLogToConsole('Occluder draw ' + (i + 1) + ' / ' + this.m_numOccluderDraws + ' : ' + 'Stencil clear');
+                bufferedLogToConsole('Stencil-clearing box drawn at ' +
+                                                '(' + scissorBoxX + ', ' + scissorBoxY + ')' +
+                                                ', width = ' + scissorBoxW + ', height = ' + scissorBoxH + '.');
+
+                gl.enable(gl.SCISSOR_TEST);
+                gl.scissor(scissorBoxX, scissorBoxY, scissorBoxW, scissorBoxH);
+                gl.clearStencil(1);
+                gl.clearColor(STENCIL_CLEAR_COLOR[0], STENCIL_CLEAR_COLOR[1], STENCIL_CLEAR_COLOR[2], STENCIL_CLEAR_COLOR[3]);
+                gl.clear(gl.COLOR_BUFFER_BIT | gl.STENCIL_BUFFER_BIT);
+                gl.disable(gl.SCISSOR_TEST);
+
+                break;
+            }
+
+            default:
+                throw new Error('Invalid occluder type: ' + type);
+        }
+    }
+
+    if (this.m_occluderTypes & OCCLUDER_SCISSOR) {
+        var scissorBoxX = this.m_rnd.getInt(-SCISSOR_OFFSET, targetW - SCISSOR_OFFSET);
+        var scissorBoxY = this.m_rnd.getInt(-SCISSOR_OFFSET, targetH - SCISSOR_OFFSET);
+        var scissorBoxW = this.m_rnd.getInt(SCISSOR_MINSIZE, targetW + SCISSOR_OFFSET);
+        var scissorBoxH = this.m_rnd.getInt(SCISSOR_MINSIZE, targetH + SCISSOR_OFFSET);
+
+        bufferedLogToConsole('Scissor box drawn at ' +
+                                        '(' + scissorBoxX + ', ' + scissorBoxY + ')' +
+                                        ', width = ' + scissorBoxW + ', height = ' + scissorBoxH + '.');
+
+        gl.enable(gl.SCISSOR_TEST);
+        gl.scissor(scissorBoxX, scissorBoxY, scissorBoxW, scissorBoxH);
+    }
+
+    var query = gl.createQuery();
+    gl.beginQuery(this.m_queryMode, query);
+
+    // Draw target primitives
+
+    gl.enable(gl.DEPTH_TEST);
+    gl.enable(gl.STENCIL_TEST);
+    gl.stencilFunc(gl.EQUAL, 0, 0xFF);
+
+    for (var i = 0; i < this.m_numTargetDraws; i++) {
+        var targetVertices = this.generateVertices(2, 2, this.m_numTargetsPerDraw, 3, this.m_rnd, this.m_targetSize, 0.4, 1); // Generate vertices for target primitives
+
+        if (targetVertices.length > 0) {
+            gl.uniform4f(colorUnif, TARGET_COLOR[0], TARGET_COLOR[1], TARGET_COLOR[2], TARGET_COLOR[3]);
+            gl.bufferData(gl.ARRAY_BUFFER, new Float32Array(targetVertices), gl.STATIC_DRAW);
+            gl.drawArrays(gl.TRIANGLES, 0, 3 * this.m_numTargetsPerDraw);
+        }
+    }
+
+    gl.endQuery(this.m_queryMode);
+    gl.finish();
+    gl.disable(gl.SCISSOR_TEST);
+    gl.disable(gl.STENCIL_TEST);
+    gl.disable(gl.DEPTH_TEST);
+
+    // Check that query result is available.
+    var resultAvailable = /** @type {boolean} */ (gl.getQueryParameter(query, gl.QUERY_RESULT_AVAILABLE));
+    assertMsgOptions(resultAvailable, 'Occlusion query failed to return a result after glFinish()', false, true);
```

Comment:
These sorts of checks in these tests -- where a query's result must be available after a glFinish -- are going to have to be restructured. See the text in this WebGL 1.0 extension:

https://www.khronos.org/registry/webgl/extensions/EXT_disjoint_timer_query/

"A query's result must not be made available until the current frame has completed rendering."

This is needed to provide uniform behavior among implementations, and prevent developers from adding busy loops to their applications.

I'm soon going to restructure Chromium's implementation of WebGL query objects to honor that requirement, at which point basically all of the ported dEQP tests that use queries are going to break. We will need to change them to issue multiple requestAnimationFrame calls, waiting for the query result to be available, and then proceed with the test.

How difficult is it going to be to add a "second part" to all of the tests which use queries? I'm thinking of something like a callback function which acts as a predicate indicating whether it can proceed, and another callback with the rest of the test. Or, multiple callbacks in the case where the test issues multiple queries.